sys/net/application_layer/sock_dns/dns_cache.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2022 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup net_sock_dns
+ * @{
+ * @file
+ * @brief   DNS cache implementation
+ * @author  Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include "checksum/fletcher32.h"
+#include "net/sock/dns.h"
+#include "dns_cache.h"
+#include "ztimer.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static struct dns_cache_entry {
+    uint32_t hash;
+    uint32_t expires;
+    union {
+#if IS_ACTIVE(SOCK_HAS_IPV4)
+        ipv4_addr_t v4;
+#endif
+#if IS_ACTIVE(SOCK_HAS_IPV6)
+        ipv6_addr_t v6;
+#endif
+    } addr;
+    uint8_t len;
+} cache[CONFIG_DNS_CACHE_SIZE];
+
+static uint8_t _addr_len(int family)
+{
+    switch (family) {
+    case AF_INET:
+        return sizeof(ipv4_addr_t);
+    case AF_INET6:
+        return sizeof(ipv6_addr_t);
+    }
+    return 0;
+}
+
+int sock_dns_cache_query(const char *domain_name, void *addr_out, int family)
+{
+    uint32_t now = ztimer_now(ZTIMER_MSEC) / MS_PER_SEC;
+    uint32_t hash = fletcher32((void *)domain_name, strlen(domain_name) / 2);
+    uint8_t addr_len = _addr_len(family);
+
+    for (unsigned i = 0; i < CONFIG_DNS_CACHE_SIZE; ++i) {
+        /* empty slot */
+        if (cache[i].len == 0) {
+            continue;
+        }
+        /* TTL expired - invalidate slot */
+        if (now > cache[i].expires) {
+            DEBUG("dns_cache[%u] expired\n", i);
+            cache[i].len = 0;
+            continue;
+        }
+        /* check if hash and length match */
+        if (cache[i].hash == hash && (!addr_len || addr_len == cache[i].len)) {
+            DEBUG("dns_cache[%u] hit\n", i);
+            memcpy(addr_out, &cache[i].addr, cache[i].len);
+            return cache[i].len;
+        }
+    }
+    DEBUG("dns_cache miss\n");
+
+    return 0;
+}
+
+static void _add_entry(uint8_t i, uint32_t hash, const void *addr_out,
+                       int addr_len, uint32_t expires)
+{
+    DEBUG("dns_cache[%u] add cache entry\n", i);
+    cache[i].hash = hash;
+    cache[i].len = addr_len;
+    cache[i].expires = expires;
+    memcpy(&cache[i].addr, addr_out, addr_len);
+}
+
+void sock_dns_cache_add(const char *domain_name, const void *addr_out,
+                        int addr_len, uint32_t ttl)
+{
+    uint32_t now = ztimer_now(ZTIMER_MSEC) / MS_PER_SEC;
+    uint32_t hash = fletcher32((void *)domain_name, strlen(domain_name) / 2);
+    uint32_t oldest = ttl;
+    int idx = -1;
+
+    DEBUG("dns_cache: lifetime of %s is %lu s\n", domain_name, ttl);
```

Comment:
On my `esp32-wroom-32` I get `RIOT/sys/net/application_layer/sock_dns/dns_cache.c:98:11: error: format '%lu' expects argument of type 'long unsigned int', but argument 3 has type 'uint32_t {aka unsigned int}' [-Werror=format=]`